ComponentKit/Utilities/CKArrayControllerChangesetVerification.mm

```diff
@@ -0,0 +1,154 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import "CKArrayControllerChangesetVerification.h"
+
+static NSMutableArray<NSNumber *> *sectionCountsForSections(NSArray<NSArray *> *sections)
+{
+  NSMutableArray *sectionCounts = [NSMutableArray new];
+  for (NSArray *section in sections) {
+    [sectionCounts addObject:@(section.count)];
+  }
+  return sectionCounts;
+}
+
+CKBadChangesetOperationType CKIsValidChangesetForSections(CKArrayControllerInputChangeset changeset, NSArray<NSArray *> *sections)
+{
+  NSMutableArray<NSNumber *> *currentSectionCounts = sectionCountsForSections(sections);
+  __block BOOL invalidChangeFound = NO;
+
+  // Updates
+  changeset.items.enumerateItems(^(NSInteger section, NSInteger index, id<NSObject> object, BOOL *stop) {
+    if (section >= [currentSectionCounts count] || index >= [currentSectionCounts[section] integerValue]) {
+      invalidChangeFound = YES;
+      *stop = YES;
+    }
+  }, nil, nil, nil);
+
+  if (invalidChangeFound) {
+    return CKBadChangesetOperationTypeUpdate;
+  }
+
+  /** 
```

Comment:
Nit: `/*` for a regular multi-line comment. No need for `/**` since this isn't a docblock.